### ISSUE TITLE
assertion failed vl_simplenn.m (see also #34)

### ISSUE THREAD

**Issue by rap9430** [vlfeat/matconvnet#45]
(This issue is similar to the #34 one)
I think i am missing something in the response (#34).Could you please elaborate on this? 
Specifically, when i change the last layer from softmaxloss->softmax in e.g. cnn_cifar example, then i get an
"Error using .* Array dimensions must match for binary array op." 
in 
"Error in vl_nnsoftmax (line 30)
Y = Y .* bsxfun(@minus, dzdY, sum(dzdY .* Y, 3)) ;"

**Comment by lenck** [comment]
Hi, sorry for such a late response. 
The issue #34 is mainly about the difference between `softmax` and `softmaxloss`. `softmaxloss` layer is in fact sort of two layers - a `softmax` with a logistic loss on top.

So, in case of the issue #34, when you want to output the class probabilities, `softmax` is the way to go, whereas when you train, you usually know the GT labels and you want to optimise some sort of loss based on the mistakes the network makes. So, then, `softmaxloss` is the friend here as it does both :)

But of course, you can also use the softmax on top even for backpropagation, but then you would have to add the loss layer on top of that... Honestly, I get lost in this quite often too (I just hope that I'm correct ;)), but for example I quite like e.g. [this](http://ufldl.stanford.edu/wiki/index.php/Softmax_Regression) explanation of softmax.
